Adding a remove-clusters command

nikhiljindal opened the pull request:
As per https://github.com/GoogleCloudPlatform/k8s-multicluster-ingress/issues/58, adding a new `kubemci remove-clusters` command that users can use to remove an existing MCI from some clusters.

Things left to be done:
* This still causes downtime. Filed https://github.com/GoogleCloudPlatform/k8s-multicluster-ingress/issues/145 for that.
* This doesnt use --force yet.
* Adding more tests.

This PR has already become too big, so will do those in follow up PRs.

cc @csbell @G-Harmon

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/googlecloudplatform/k8s-multicluster-ingress/146)
<!-- Reviewable:end -->

nikhiljindal commented:
Pushed a new commit to add a few more unit tests

G-Harmon commented:
Reviewed 16 of 17 files at r1, 2 of 2 files at r2.
Review status: all files reviewed at latest revision, 17 unresolved discussions.

---

*[app/kubemci/cmd/remove_clusters.go, line 59 at r2](https://reviewable.io:443/reviews/googlecloudplatform/k8s-multicluster-ingress/146#-L7kBah3jAL8d_pJVOGm:-L7kBah3jAL8d_pJVOGn:b-qdx0ra) ([raw file](https://github.com/googlecloudplatform/k8s-multicluster-ingress/blob/50c253a4797a2a66d9d6c2f7bd7f3234c60ee01b/app/kubemci/cmd/remove_clusters.go#L59)):*
> // Overwrite values when they differ from what's requested. If
> 	// the resource does not exist, or is already the correct
> 	// value, then 'force' is a no-op.
> 	ForceUpdate bool
Is "Force" meaningful for remove-clusters?

---

*[app/kubemci/cmd/remove_clusters.go, line 62 at r2](https://reviewable.io:443/reviews/googlecloudplatform/k8s-multicluster-ingress/146#-L7kBoq-RvY2oJfE_ZBu:-L7kBoq-RvY2oJfE_ZBv:b-9ghe27) ([raw file](https://github.com/googlecloudplatform/k8s-multicluster-ingress/blob/50c253a4797a2a66d9d6c2f7bd7f3234c60ee01b/app/kubemci/cmd/remove_clusters.go#L62)):*
> // Optional.
It's inconsistent that some members say "required", some say "optional", and some don't say either....

---

*[app/kubemci/cmd/remove_clusters.go, line 101 at r2](https://reviewable.io:443/reviews/googlecloudplatform/k8s-multicluster-ingress/146#-L7kC-OwZCTCjl87VaRe:-L7kC-OwZCTCjl87VaRf:bs6puzg) ([raw file](https://github.com/googlecloudplatform/k8s-multicluster-ingress/blob/50c253a4797a2a66d9d6c2f7bd7f3234c60ee01b/app/kubemci/cmd/remove_clusters.go#L101)):*
> params
nit: s/params/options/ ? or s/params/flags/ ?

---

*[app/kubemci/cmd/remove_clusters.go, line 148 at r2](https://reviewable.io:443/reviews/googlecloudplatform/k8s-multicluster-ingress/146#-L7kD8W1p_4PKHIQzjsB:-L7kD8W1p_4PKHIQzjsC:b-9queov) ([raw file](https://github.com/googlecloudplatform/k8s-multicluster-ingress/blob/50c253a4797a2a66d9d6c2f7bd7f3234c60ee01b/app/kubemci/cmd/remove_clusters.go#L148)):*
> ```Go
> 
> 	// Delete ingress resource from clusters
> 	err = ingress.NewIngressSyncer().DeleteIngress(&ing, clients)
> ```

shouldn't we check NewIngressSyncer() for returning nil? (I know it doesn't know, but isn't the pattern to check for nil from a New* func?)

---

*[app/kubemci/pkg/gcp/backendservice/backendservicesyncer.go, line 273 at r2](https://reviewable.io:443/reviews/googlecloudplatform/k8s-multicluster-ingress/146#-L7kG4Vr5tG3M-6461Dj:-L7kG4Vr5tG3M-6461Dk:b-yt4nf) ([raw file](https://github.com/googlecloudplatform/k8s-multicluster-ingress/blob/50c253a4797a2a66d9d6c2f7bd7f3234c60ee01b/app/kubemci/pkg/gcp/backendservice/backendservicesyncer.go#L273)):*
> ```Go
> }
> 
> func (b *BackendServiceSyncer) desiredBackendServiceWithoutClusters(existingBE *compute.BackendService, removeIGLinks []string) *compute.BackendService {
> ```

minor: can you add a comment please for my benefit?

---

*[app/kubemci/pkg/gcp/backendservice/backendservicesyncer_test.go, line 221 at r2](https://reviewable.io:443/reviews/googlecloudplatform/k8s-multicluster-ingress/146#-L7kH2epoLl5PetsqC7S:-L7kH2epoLl5PetsqC7T:b-nwubv3) ([raw file](https://github.com/googlecloudplatform/k8s-multicluster-ingress/blob/50c253a4797a2a66d9d6c2f7bd7f3234c60ee01b/app/kubemci/pkg/gcp/backendservice/backendservicesyncer_test.go#L221)):*
> if len(be.Backends) != 1 {
> 		t.Fatalf("expected the backend service to have 1 backend for ig: %s, actual: %v", ig2Link, be.Backends)
> 	}
>  
Can you please verify that the right ig is left?

---

*[app/kubemci/pkg/gcp/backendservice/fake_backendservicesyncer.go, line 66 at r2](https://reviewable.io:443/reviews/googlecloudplatform/k8s-multicluster-ingress/146#-L7kFJUukLyJVyy3wzzU:-L7kFJUukLyJVyy3wzzV:b-cfuhn1) ([raw file](https://github.com/googlecloudplatform/k8s-multicluster-ingress/blob/50c253a4797a2a66d9d6c2f7bd7f3234c60ee01b/app/kubemci/pkg/gcp/backendservice/fake_backendservicesyncer.go#L66)):*
> ```Go
> 
> func (h *FakeBackendServiceSyncer) RemoveFromClusters(ports []ingressbe.ServicePort, removeIGLinks []string) error {
> 	// TODO: Implement this.
> ```

put in a glog.Fatal? Would be better for folks to know that this doesn't work if they try it.

---

*[app/kubemci/pkg/gcp/firewallrule/firewallrulesyncer.go, line 101 at r2](https://reviewable.io:443/reviews/googlecloudplatform/k8s-multicluster-ingress/146#-L7kI2YR4HmaBla3mC9r:-L7kI2YR4HmaBla3mC9s:b-y1pcqs) ([raw file](https://github.com/googlecloudplatform/k8s-multicluster-ingress/blob/50c253a4797a2a66d9d6c2f7bd7f3234c60ee01b/app/kubemci/pkg/gcp/firewallrule/firewallrulesyncer.go#L101)):*
> func (s *FirewallRuleSyncer) removeFromClusters(lbName string, removeIGLinks map[string][]string) error {
I don't understand why you have both RemoveFromClusters and the helper removeFromClusters. In this case, the functions have the same signature and there's loop in the outer func.

---

*[app/kubemci/pkg/gcp/firewallrule/firewallrulesyncer.go, line 257 at r2](https://reviewable.io:443/reviews/googlecloudplatform/k8s-multicluster-ingress/146#-L7kIU2oq3tSESvjbNH0:-L7kIU2oq3tSESvjbNH1:b-61nq6f) ([raw file](https://github.com/googlecloudplatform/k8s-multicluster-ingress/blob/50c253a4797a2a66d9d6c2f7bd7f3234c60ee01b/app/kubemci/pkg/gcp/firewallrule/firewallrulesyncer.go#L257)):*
> // TODO(nikhiljindal): Fix this if it becomes a problem.
File an Issue so we can point to a known issue if necessary?

---

*[app/kubemci/pkg/gcp/forwardingrule/forwardingrulesyncer.go, line 212 at r2](https://reviewable.io:443/reviews/googlecloudplatform/k8s-multicluster-ingress/146#-L7kLkwbxerlB7HQfWbI:-L7kLkwbxerlB7HQfWbJ:b-x1mz56) ([raw file](https://github.com/googlecloudplatform/k8s-multicluster-ingress/blob/50c253a4797a2a66d9d6c2f7bd7f3234c60ee01b/app/kubemci/pkg/gcp/forwardingrule/forwardingrulesyncer.go#L212)):*
> // Update HTTPS status first and then HTTP.
> 	// This ensures that get-status returns the correct status even if HTTPS is updated but updating HTTP fails.
> 	// This is because get-status reads from HTTP if it exists.
> 	// Also note that it continues silently if either HTTP or HTTPS forwarding rules do not exist.
I don't quite follow all this logic, but might not be a big deal.

---

*[app/kubemci/pkg/gcp/forwardingrule/forwardingrulesyncer.go, line 219 at r2](https://reviewable.io:443/reviews/googlecloudplatform/k8s-multicluster-ingress/146#-L7kKb0qk9msdnnqIJpp:-L7kKb0qk9msdnnqIJpq:b-34jqfs) ([raw file](https://github.com/googlecloudplatform/k8s-multicluster-ingress/blob/50c253a4797a2a66d9d6c2f7bd7f3234c60ee01b/app/kubemci/pkg/gcp/forwardingrule/forwardingrulesyncer.go#L219)):*
> ```Go
> }
> 
> // ensureForwardingRule ensures a forwarding rule exists as per the given input parameters.
> ```

copy-pasted comment needs updating.

---

*[app/kubemci/pkg/gcp/instances/fake_instancegetter.go, line 24 at r2](https://reviewable.io:443/reviews/googlecloudplatform/k8s-multicluster-ingress/146#-L7kJJsExHGRc-ewvtQ8:-L7kJJsExHGRc-ewvtQ9:b-kn7gik) ([raw file](https://github.com/googlecloudplatform/k8s-multicluster-ingress/blob/50c253a4797a2a66d9d6c2f7bd7f3234c60ee01b/app/kubemci/pkg/gcp/instances/fake_instancegetter.go#L24)):*
> Zone: igUrl,
seems fishy to use the same URL for Name and Zone and Tag.

---

*[app/kubemci/pkg/gcp/loadbalancer/loadbalancersyncer.go, line 138 at r2](https://reviewable.io:443/reviews/googlecloudplatform/k8s-multicluster-ingress/146#-L7kESq6eywSBYwR7Sd9:-L7kESq6eywSBYwR7SdA:b-tnzdg) ([raw file](https://github.com/googlecloudplatform/k8s-multicluster-ingress/blob/50c253a4797a2a66d9d6c2f7bd7f3234c60ee01b/app/kubemci/pkg/gcp/loadbalancer/loadbalancersyncer.go#L138)):*
>  portsand 
s/portsand/ports and/

---

*[app/kubemci/pkg/gcp/loadbalancer/loadbalancersyncer.go, line 266 at r2](https://reviewable.io:443/reviews/googlecloudplatform/k8s-multicluster-ingress/146#-L7kDWRzivcUVMKCnn8b:-L7kDWRzivcUVMKCnn8c:b970ao8) ([raw file](https://github.com/googlecloudplatform/k8s-multicluster-ingress/blob/50c253a4797a2a66d9d6c2f7bd7f3234c60ee01b/app/kubemci/pkg/gcp/loadbalancer/loadbalancersyncer.go#L266)):*
> ```Go
> // DeleteLoadBalancer deletes the GCP resources associated with the L7 GCP load balancer for the given ingress.
> func (l *LoadBalancerSyncer) RemoveFromClusters(ing *v1beta1.Ingress, removeClients map[string]kubeclient.Interface, forceUpdate bool) error {
> 	// TODO(nikhiljindal): Dont require the ingress yaml from users. Just the name should be enough. We can fetch ingress YAML from one of the clusters.
> ```

Why is this TODO placed right here? I would either put this in the function comment, or next to ing's first use.

---

*[app/kubemci/pkg/gcp/loadbalancer/loadbalancersyncer.go, line 291 at r2](https://reviewable.io:443/reviews/googlecloudplatform/k8s-multicluster-ingress/146#-L7kE4w98zMLWivNmZe0:-L7kE4w98zMLWivNmZe1:b-6y75od) ([raw file](https://github.com/googlecloudplatform/k8s-multicluster-ingress/blob/50c253a4797a2a66d9d6c2f7bd7f3234c60ee01b/app/kubemci/pkg/gcp/loadbalancer/loadbalancersyncer.go#L291)):*
> igsForBE := []string{}
> 	for k := range removeIGLinks {
> 		igsForBE = append(igsForBE, removeIGLinks[k]...)
> 	}
What's this doing?

---

*[app/kubemci/pkg/gcp/loadbalancer/loadbalancersyncer.go, line 383 at r2](https://reviewable.io:443/reviews/googlecloudplatform/k8s-multicluster-ingress/146#-L7kEiw0I-d5IP6Wgjmi:-L7kEiw0I-d5IP6Wgjmj:b96hqff) ([raw file](https://github.com/googlecloudplatform/k8s-multicluster-ingress/blob/50c253a4797a2a66d9d6c2f7bd7f3234c60ee01b/app/kubemci/pkg/gcp/loadbalancer/loadbalancersyncer.go#L383)):*
> // Also note that this returns all named ports on the instance group and not just the ones relevant to the given ingress.
Is this a problem? or, what problem might this cause?

---

*[app/kubemci/pkg/gcp/loadbalancer/loadbalancersyncer_test.go, line 366 at r2](https://reviewable.io:443/reviews/googlecloudplatform/k8s-multicluster-ingress/146#-L7kJgfe3u9iQ4QHcGaq:-L7kJgfe3u9iQ4QHcGar:bucbzax) ([raw file](https://github.com/googlecloudplatform/k8s-multicluster-ingress/blob/50c253a4797a2a66d9d6c2f7bd7f3234c60ee01b/app/kubemci/pkg/gcp/loadbalancer/loadbalancersyncer_test.go#L366)):*
> ```Go
> 	ing, err := setupLBCForCreateIng(lbc, nodePort, igName, igZone, zoneLink, ipAddress)
> 	if err != nil {
> 		t.Fatalf("%s", err)
> ```

nit: include some text besides the error

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/googlecloudplatform/k8s-multicluster-ingress/146)*
<!-- Sent from Reviewable.io -->

nikhiljindal commented:
Review status: all files reviewed at latest revision, 17 unresolved discussions.

---

*[app/kubemci/cmd/remove_clusters.go, line 59 at r2](https://reviewable.io:443/reviews/googlecloudplatform/k8s-multicluster-ingress/146#-L7kBah3jAL8d_pJVOGm:-L80dqeeMUFTUu40jmF5:b-sw1oz4) ([raw file](https://github.com/googlecloudplatform/k8s-multicluster-ingress/blob/50c253a4797a2a66d9d6c2f7bd7f3234c60ee01b/app/kubemci/cmd/remove_clusters.go#L59)):*
<details><summary><i>Previously, G-Harmon wrote…</i></summary><blockquote>

> // Overwrite values when they differ from what's requested. If
> 	// the resource does not exist, or is already the correct
> 	// value, then 'force' is a no-op.
> 	ForceUpdate bool
Is "Force" meaningful for remove-clusters?
</blockquote></details>

https://github.com/GoogleCloudPlatform/k8s-multicluster-ingress/issues/58 has an idea of what we can use it for.

---

*[app/kubemci/cmd/remove_clusters.go, line 62 at r2](https://reviewable.io:443/reviews/googlecloudplatform/k8s-multicluster-ingress/146#-L7kBoq-RvY2oJfE_ZBu:-L80ewuXXY8OfqR0KxFj:b-hygti4) ([raw file](https://github.com/googlecloudplatform/k8s-multicluster-ingress/blob/50c253a4797a2a66d9d6c2f7bd7f3234c60ee01b/app/kubemci/cmd/remove_clusters.go#L62)):*
<details><summary><i>Previously, G-Harmon wrote…</i></summary><blockquote>

> // Optional.
It's inconsistent that some members say "required", some say "optional", and some don't say either....
</blockquote></details>

Good catch. Fixed so that required params explicitly say Required.

---

*[app/kubemci/cmd/remove_clusters.go, line 101 at r2](https://reviewable.io:443/reviews/googlecloudplatform/k8s-multicluster-ingress/146#-L7kC-OwZCTCjl87VaRe:-L80fd8yYG2Ia4wPv3Tq:b-896fix) ([raw file](https://github.com/googlecloudplatform/k8s-multicluster-ingress/blob/50c253a4797a2a66d9d6c2f7bd7f3234c60ee01b/app/kubemci/cmd/remove_clusters.go#L101)):*
<details><summary><i>Previously, G-Harmon wrote…</i></summary><blockquote>

> params
nit: s/params/options/ ? or s/params/flags/ ?
</blockquote></details>

Done.

---

*[app/kubemci/cmd/remove_clusters.go, line 148 at r2](https://reviewable.io:443/reviews/googlecloudplatform/k8s-multicluster-ingress/146#-L7kD8W1p_4PKHIQzjsB:-L80je8GFudTCxSI-rGk:b-896fix) ([raw file](https://github.com/googlecloudplatform/k8s-multicluster-ingress/blob/50c253a4797a2a66d9d6c2f7bd7f3234c60ee01b/app/kubemci/cmd/remove_clusters.go#L148)):*
<details><summary><i>Previously, G-Harmon wrote…</i></summary><blockquote>

shouldn't we check NewIngressSyncer() for returning nil? (I know it doesn't know, but isn't the pattern to check for nil from a New* func?)
</blockquote></details>

Done.

---

*[app/kubemci/pkg/gcp/backendservice/backendservicesyncer.go, line 273 at r2](https://reviewable.io:443/reviews/googlecloudplatform/k8s-multicluster-ingress/146#-L7kG4Vr5tG3M-6461Dj:-L80kNBAGn9ZGWMWjljo:b4jchf4) ([raw file](https://github.com/googlecloudplatform/k8s-multicluster-ingress/blob/50c253a4797a2a66d9d6c2f7bd7f3234c60ee01b/app/kubemci/pkg/gcp/backendservice/backendservicesyncer.go#L273)):*
<details><summary><i>Previously, G-Harmon wrote…</i></summary><blockquote>

minor: can you add a comment please for my benefit?
</blockquote></details>

Sure, done

---

*[app/kubemci/pkg/gcp/backendservice/backendservicesyncer_test.go, line 221 at r2](https://reviewable.io:443/reviews/googlecloudplatform/k8s-multicluster-ingress/146#-L7kH2epoLl5PetsqC7S:-L80lrteYgA3vc4v5xGC:b-896fix) ([raw file](https://github.com/googlecloudplatform/k8s-multicluster-ingress/blob/50c253a4797a2a66d9d6c2f7bd7f3234c60ee01b/app/kubemci/pkg/gcp/backendservice/backendservicesyncer_test.go#L221)):*
<details><summary><i>Previously, G-Harmon wrote…</i></summary><blockquote>

> if len(be.Backends) != 1 {
> 		t.Fatalf("expected the backend service to have 1 backend for ig: %s, actual: %v", ig2Link, be.Backends)
> 	}
>  
Can you please verify that the right ig is left?
</blockquote></details>

Done.

---

*[app/kubemci/pkg/gcp/backendservice/fake_backendservicesyncer.go, line 66 at r2](https://reviewable.io:443/reviews/googlecloudplatform/k8s-multicluster-ingress/146#-L7kFJUukLyJVyy3wzzU:-L80m9Aa9NlNaK6EbiMD:b-s0vpwf) ([raw file](https://github.com/googlecloudplatform/k8s-multicluster-ingress/blob/50c253a4797a2a66d9d6c2f7bd7f3234c60ee01b/app/kubemci/pkg/gcp/backendservice/fake_backendservicesyncer.go#L66)):*
<details><summary><i>Previously, G-Harmon wrote…</i></summary><blockquote>

put in a glog.Fatal? Would be better for folks to know that this doesn't work if they try it.
</blockquote></details>

Implemented it :)

---

*[app/kubemci/pkg/gcp/firewallrule/firewallrulesyncer.go, line 101 at r2](https://reviewable.io:443/reviews/googlecloudplatform/k8s-multicluster-ingress/146#-L7kI2YR4HmaBla3mC9r:-L83eYxRqHSB7FXtvyYI:b-22w2jj) ([raw file](https://github.com/googlecloudplatform/k8s-multicluster-ingress/blob/50c253a4797a2a66d9d6c2f7bd7f3234c60ee01b/app/kubemci/pkg/gcp/firewallrule/firewallrulesyncer.go#L101)):*
<details><summary><i>Previously, G-Harmon wrote…</i></summary><blockquote>

> func (s *FirewallRuleSyncer) removeFromClusters(lbName string, removeIGLinks map[string][]string) error {
I don't understand why you have both RemoveFromClusters and the helper removeFromClusters. In this case, the functions have the same signature and there's loop in the outer func.
</blockquote></details>

Makes it easier to prefix "error in removing clusters from firewall rule" to all errors.

---

*[app/kubemci/pkg/gcp/firewallrule/firewallrulesyncer.go, line 257 at r2](https://reviewable.io:443/reviews/googlecloudplatform/k8s-multicluster-ingress/146#-L7kIU2oq3tSESvjbNH0:-L83sGuBoON209bg9lGk:beijdpu) ([raw file](https://github.com/googlecloudplatform/k8s-multicluster-ingress/blob/50c253a4797a2a66d9d6c2f7bd7f3234c60ee01b/app/kubemci/pkg/gcp/firewallrule/firewallrulesyncer.go#L257)):*
> This assumes that the ordering of target tags has not changed on these instances.
> 	// A potential solution to that problem is to recompute the target tags for the existing clusters,
> 	// rather than removing the ones for old clusters.
> 	// But we do not want to change existing tags for clusters that are already working.
> 	// Ideally network tags should only be appended to, so this should not be a problem.
> 	// TODO(nikhiljindal): Fix this if it becomes a problem.

Good idea. Filed https://github.com/GoogleCloudPlatform/k8s-multicluster-ingress/issues/147 and added link to comments in code.

---

*[app/kubemci/pkg/gcp/forwardingrule/forwardingrulesyncer.go, line 212 at r2](https://reviewable.io:443/reviews/googlecloudplatform/k8s-multicluster-ingress/146#-L7kLkwbxerlB7HQfWbI:-L83t6hid2Ml1XgKeyPe:b-tta7ct) ([raw file](https://github.com/googlecloudplatform/k8s-multicluster-ingress/blob/50c253a4797a2a66d9d6c2f7bd7f3234c60ee01b/app/kubemci/pkg/gcp/forwardingrule/forwardingrulesyncer.go#L212)):*
<details><summary><i>Previously, G-Harmon wrote…</i></summary><blockquote>

> // Update HTTPS status first and then HTTP.
> 	// This ensures that get-status returns the correct status even if HTTPS is updated but updating HTTP fails.
> 	// This is because get-status reads from HTTP if it exists.
> 	// Also note that it continues silently if either HTTP or HTTPS forwarding rules do not exist.
I don't quite follow all this logic, but might not be a big deal.
</blockquote></details>

Reworded. Hopefully its better now

---

*[app/kubemci/pkg/gcp/forwardingrule/forwardingrulesyncer.go, line 219 at r2](https://reviewable.io:443/reviews/googlecloudplatform/k8s-multicluster-ingress/146#-L7kKb0qk9msdnnqIJpp:-L83tQG2vqwLAW7IyHcg:b-896fix) ([raw file](https://github.com/googlecloudplatform/k8s-multicluster-ingress/blob/50c253a4797a2a66d9d6c2f7bd7f3234c60ee01b/app/kubemci/pkg/gcp/forwardingrule/forwardingrulesyncer.go#L219)):*
<details><summary><i>Previously, G-Harmon wrote…</i></summary><blockquote>

copy-pasted comment needs updating.
</blockquote></details>

Done.

---

*[app/kubemci/pkg/gcp/instances/fake_instancegetter.go, line 24 at r2](https://reviewable.io:443/reviews/googlecloudplatform/k8s-multicluster-ingress/146#-L7kJJsExHGRc-ewvtQ8:-L83tnIj0uOytEilBDB9:bv38i3i) ([raw file](https://github.com/googlecloudplatform/k8s-multicluster-ingress/blob/50c253a4797a2a66d9d6c2f7bd7f3234c60ee01b/app/kubemci/pkg/gcp/instances/fake_instancegetter.go#L24)):*
<details><summary><i>Previously, G-Harmon wrote…</i></summary><blockquote>

> Zone: igUrl,
seems fishy to use the same URL for Name and Zone and Tag.
</blockquote></details>

Its a fake :)

---

*[app/kubemci/pkg/gcp/loadbalancer/loadbalancersyncer.go, line 138 at r2](https://reviewable.io:443/reviews/googlecloudplatform/k8s-multicluster-ingress/146#-L7kESq6eywSBYwR7Sd9:-L83ttDhJ3hujDBb0KoF:b-896fix) ([raw file](https://github.com/googlecloudplatform/k8s-multicluster-ingress/blob/50c253a4797a2a66d9d6c2f7bd7f3234c60ee01b/app/kubemci/pkg/gcp/loadbalancer/loadbalancersyncer.go#L138)):*
<details><summary><i>Previously, G-Harmon wrote…</i></summary><blockquote>

>  portsand 
s/portsand/ports and/
</blockquote></details>

Done.

---

*[app/kubemci/pkg/gcp/loadbalancer/loadbalancersyncer.go, line 266 at r2](https://reviewable.io:443/reviews/googlecloudplatform/k8s-multicluster-ingress/146#-L7kDWRzivcUVMKCnn8b:-L83u7bKyo3vlp-zPvZv:b-d4vst9) ([raw file](https://github.com/googlecloudplatform/k8s-multicluster-ingress/blob/50c253a4797a2a66d9d6c2f7bd7f3234c60ee01b/app/kubemci/pkg/gcp/loadbalancer/loadbalancersyncer.go#L266)):*
<details><summary><i>Previously, G-Harmon wrote…</i></summary><blockquote>

Why is this TODO placed right here? I would either put this in the function comment, or next to ing's first use.
</blockquote></details>

Good point. Moved to function comment. Did the same for DeleteLoadBalancer, which had the same comment.

---

*[app/kubemci/pkg/gcp/loadbalancer/loadbalancersyncer.go, line 291 at r2](https://reviewable.io:443/reviews/googlecloudplatform/k8s-multicluster-ingress/146#-L7kE4w98zMLWivNmZe0:-L83uUAWmWnTIFKSb0rJ:b-5wlrnk) ([raw file](https://github.com/googlecloudplatform/k8s-multicluster-ingress/blob/50c253a4797a2a66d9d6c2f7bd7f3234c60ee01b/app/kubemci/pkg/gcp/loadbalancer/loadbalancersyncer.go#L291)):*
<details><summary><i>Previously, G-Harmon wrote…</i></summary><blockquote>

> igsForBE := []string{}
> 	for k := range removeIGLinks {
> 		igsForBE = append(igsForBE, removeIGLinks[k]...)
> 	}
What's this doing?
</blockquote></details>

This is converting a map of clustername to IGs for that cluster into a flat array of all IGs.
Added comment in code.

---

*[app/kubemci/pkg/gcp/loadbalancer/loadbalancersyncer.go, line 383 at r2](https://reviewable.io:443/reviews/googlecloudplatform/k8s-multicluster-ingress/146#-L7kEiw0I-d5IP6Wgjmi:-L83uodWvkEebDm3P4xI:bhobeeh) ([raw file](https://github.com/googlecloudplatform/k8s-multicluster-ingress/blob/50c253a4797a2a66d9d6c2f7bd7f3234c60ee01b/app/kubemci/pkg/gcp/loadbalancer/loadbalancersyncer.go#L383)):*
<details><summary><i>Previously, G-Harmon wrote…</i></summary><blockquote>

> // Also note that this returns all named ports on the instance group and not just the ones relevant to the given ingress.
Is this a problem? or, what problem might this cause?
</blockquote></details>

That depends on the caller :)
This method clarifies that it returns all the ports.

---

*[app/kubemci/pkg/gcp/loadbalancer/loadbalancersyncer_test.go, line 366 at r2](https://reviewable.io:443/reviews/googlecloudplatform/k8s-multicluster-ingress/146#-L7kJgfe3u9iQ4QHcGaq:-L83vF8x3tokkqtbNzAt:b-4sukh8) ([raw file](https://github.com/googlecloudplatform/k8s-multicluster-ingress/blob/50c253a4797a2a66d9d6c2f7bd7f3234c60ee01b/app/kubemci/pkg/gcp/loadbalancer/loadbalancersyncer_test.go#L366)):*
<details><summary><i>Previously, G-Harmon wrote…</i></summary><blockquote>

nit: include some text besides the error
</blockquote></details>

Done at all places in this test.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/googlecloudplatform/k8s-multicluster-ingress/146)*
<!-- Sent from Reviewable.io -->

nikhiljindal commented:
Thanks for the detailed review @G-Harmon 
Updated as per comments. PTAL.

G-Harmon commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 9 of 9 files at r3.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[app/kubemci/pkg/gcp/backendservice/fake_backendservicesyncer.go, line 66 at r2](https://reviewable.io:443/reviews/googlecloudplatform/k8s-multicluster-ingress/146#-L7kFJUukLyJVyy3wzzU:-L88iwM33V4RgpLLbSzn:b-7m74ox) ([raw file](https://github.com/googlecloudplatform/k8s-multicluster-ingress/blob/50c253a4797a2a66d9d6c2f7bd7f3234c60ee01b/app/kubemci/pkg/gcp/backendservice/fake_backendservicesyncer.go#L66)):*
<details><summary><i>Previously, nikhiljindal (Nikhil Jindal) wrote…</i></summary><blockquote>

Implemented it :)
</blockquote></details>

nice!

---

*[app/kubemci/pkg/gcp/firewallrule/firewallrulesyncer.go, line 101 at r2](https://reviewable.io:443/reviews/googlecloudplatform/k8s-multicluster-ingress/146#-L7kI2YR4HmaBla3mC9r:-L88j9D7_UvCDdNmHrI2:b-wwwgl5) ([raw file](https://github.com/googlecloudplatform/k8s-multicluster-ingress/blob/50c253a4797a2a66d9d6c2f7bd7f3234c60ee01b/app/kubemci/pkg/gcp/firewallrule/firewallrulesyncer.go#L101)):*
<details><summary><i>Previously, nikhiljindal (Nikhil Jindal) wrote…</i></summary><blockquote>

Makes it easier to prefix "error in removing clusters from firewall rule" to all errors.
</blockquote></details>

haha, ok.

---

*[app/kubemci/pkg/gcp/firewallrule/firewallrulesyncer.go, line 257 at r2](https://reviewable.io:443/reviews/googlecloudplatform/k8s-multicluster-ingress/146#-L7kIU2oq3tSESvjbNH0:-L88k1QA3rhzivRyb64O:b-k73z76) ([raw file](https://github.com/googlecloudplatform/k8s-multicluster-ingress/blob/50c253a4797a2a66d9d6c2f7bd7f3234c60ee01b/app/kubemci/pkg/gcp/firewallrule/firewallrulesyncer.go#L257)):*
<details><summary><i>Previously, nikhiljindal (Nikhil Jindal) wrote…</i></summary><blockquote>

> This assumes that the ordering of target tags has not changed on these instances.
> 	// A potential solution to that problem is to recompute the target tags for the existing clusters,
> 	// rather than removing the ones for old clusters.
> 	// But we do not want to change existing tags for clusters that are already working.
> 	// Ideally network tags should only be appended to, so this should not be a problem.
> 	// TODO(nikhiljindal): Fix this if it becomes a problem.

Good idea. Filed https://github.com/GoogleCloudPlatform/k8s-multicluster-ingress/issues/147 and added link to comments in code.
</blockquote></details>

thanks

---

*[app/kubemci/pkg/gcp/forwardingrule/forwardingrulesyncer.go, line 212 at r2](https://reviewable.io:443/reviews/googlecloudplatform/k8s-multicluster-ingress/146#-L7kLkwbxerlB7HQfWbI:-L88jNy2rYhOyRjMTFR2:b-8cj406) ([raw file](https://github.com/googlecloudplatform/k8s-multicluster-ingress/blob/50c253a4797a2a66d9d6c2f7bd7f3234c60ee01b/app/kubemci/pkg/gcp/forwardingrule/forwardingrulesyncer.go#L212)):*
<details><summary><i>Previously, nikhiljindal (Nikhil Jindal) wrote…</i></summary><blockquote>

Reworded. Hopefully its better now
</blockquote></details>

yes, i follow now.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/googlecloudplatform/k8s-multicluster-ingress/146#-:-L88k6xNAa-sE24Eg9VB:bnfp4nl)*
<!-- Sent from Reviewable.io -->
